src/invoice/verification.rs

```diff
@@ -0,0 +1,389 @@
+    /// is known and the signature is valid.
+    AuthoritativeIntegrity,
+    /// Verify that the Creator key is known and that all signatures are valid.
+    ///
+    /// This is subject to a DOS attack if a signer can generate intentionally bad signatures.
+    GreedyVerification,
+    /// Verify that every key on the invoice is known, and that every signature is valid.
+    ExhaustiveVerification,
+    /// Verifies that all signatures of the given roles are valid and signed by know keys.
+    ///
+    /// If the bool is true, unknown signers will also be valdidated.
+    /// Be aware that doing so may make the validation subject to a special for of
+    /// DOS attack in which someone can generate a known-bad signature.
+    MultipleAttestation(Vec<SignatureRole>, bool),
+}
+
+impl Default for VerificationStrategy {
+    fn default() -> Self {
+        VerificationStrategy::GreedyVerification
+    }
+}
+
+/// A strategy for verifying an invoice.
+impl VerificationStrategy {
+    fn verify_signature(&self, sig: &Signature, cleartext: &[u8]) -> Result<(), SignatureError> {
+        let pk = base64::decode(sig.key.as_bytes())
+            .map_err(|_| SignatureError::CorruptKey(sig.key.clone()))?;
+        let sig_block = base64::decode(sig.signature.as_bytes())
+            .map_err(|_| SignatureError::CorruptSignature(sig.key.clone()))?;
+
+        let pubkey =
+            PublicKey::from_bytes(&pk).map_err(|_| SignatureError::CorruptKey(sig.key.clone()))?;
+        let ed_sig = EdSignature::new(
+            sig_block
+                .as_slice()
+                .try_into()
+                .map_err(|_| SignatureError::CorruptSignature(sig.key.clone()))?,
+        );
+        pubkey
+            .verify_strict(cleartext, &ed_sig)
+            .map_err(|_| SignatureError::Unverified(sig.key.clone()))
+    }
+    /// Verify that every signature on this invoice is correct.
+    ///
+    /// The verification strategy will determine how this verification is performed.
+    /// Depending on the selected strategy, the `[[signature]]` blocks will be evaluated
+    /// for the following:
+    ///
+    /// - Is the key in the keyring?
+    /// - Can the signature be verified?
+    ///
+    /// Note that the purpose of the keyring is to ensure that we know about the
+    /// entity that claims to have signed the invoice.
+    ///
+    /// If no signatures are on the invoice, this will succeed.
+    ///
+    /// A strategy will determine success or failure based on whether the signature is verified,
+    /// whether the keys are known, whether the requisite number/roles are satisfied, and
+    /// so on.
+    pub fn verify(&self, inv: &Invoice, keyring: &KeyRing) -> Result<(), SignatureError> {
+        let (roles, all_valid, all_verified, all_roles) = match self {
+            VerificationStrategy::GreedyVerification => {
+                (vec![SignatureRole::Creator], true, true, true)
```

Comment:
Nit: Based on what I am seeing in the code below, I don't think we need to reallocate owned `Vec`s every time we verify. Not sure if we can do this with enum variants, but it might be worth looking into making each of these sets a `const` slice and then using those in the logic

Reply:
You mentioned you were gonna change this. So I will leave it as-is